content/blog/autoscaling-custom-k8s-metric/index.md

```diff
@@ -0,0 +1,243 @@
+
+```yaml
+apiVersion: autoscaling/v2beta2
+kind: HorizontalPodAutoscaler
+metadata:
+  name: echo-service-hpa
+spec:
+  scaleTargetRef:
+    apiVersion: apps/v1
+    kind: Deployment
+    name: echo-service
+  minReplicas: 1
+  maxReplicas: 10
+  metrics:
+    - type: Pods
+      pods:
+        metric:
+          name: px-http-requests-per-second
+        target:
+          type: AverageValue
+          averageValue: 20
+```
+
+**This autoscaler will try to meet the following goal: 20 requests per second (on average), per pod.** If there are more requests per second, it will increase the number of pods, and if there are fewer, it will decrease the number of pods. The `maxReplicas` property prevents the autoscaler from provisioning more than 10 pods.
+
+We can use [hey](https://github.com/rakyll/hey) to generate artificial HTTP load. This generates 10,000 requests to the `/ping` endpoint in our server.
+
+```bash
+hey -n 10000 http://<custom-metric-server-external-ip>/ping
+```
+
+Let's see what happens to the number of pods over time in our service. In the chart below, the 10,000 requests were generated at second 0. The requests were fast -- all 10,000 completed within a few seconds.
+
+::: div image-xl
+<svg title="The number of pods increases in response to load, then decreases." src='autoscaling-chart-light.png' />
+:::
+
+This is pretty cool. While all of the requests completed within 5 seconds, it took about 100 seconds for the autoscaler to max out the number of pods.
+
+This is because **Kubernetes autoscaling has intentional delays by default**, so that the system doesn't scale up or down too quickly due to spurious load. However, we can configure the `scaleDown`/`scaleUp` behavior to make it add or remove pods more quickly if we want:
+
+```yaml
+apiVersion: autoscaling/v2beta2
+kind: HorizontalPodAutoscaler
+metadata:
+  name: echo-service-hpa
+spec:
+  scaleTargetRef:
+    apiVersion: apps/v1
+    kind: Deployment
+    name: echo-service
+  minReplicas: 1
+  maxReplicas: 10
+  behavior:
+    scaleDown:
+      stabilizationWindowSeconds: 0
+    scaleUp:
+      stabilizationWindowSeconds: 0
+  metrics:
+    - type: Pods
+      pods:
+        metric:
+          name: px-http-requests-per-second
+        target:
+          type: AverageValue
+          averageValue: 20
+```
+
+This should increase the responsiveness of the autoscaler. Note: there are other parameters that you can configure in `behavior`, some of which are listed [here](https://kubernetes.io/docs/tasks/run-application/horizontal-pod-autoscale/#support-for-configurable-scaling-behavior).
+
+Let's compare the new autoscaling configuration to the old one applied with the exact same input stimulus: 10,000 pings from `hey` at second 0.
+
+::: div image-xl
+<svg title="The number of pods increases/decreases faster when we reduce the stabilization window." src='autoscaling-chart-fast-and-slow-light.png' />
```

Comment:
ditto drop the `-light`

(and copy `autoscaling-chart-fast-and-slow-light.png` file to `autoscaling-chart-fast-and-slow.png`)